modules/skill-email/src/actions/send_email.js

```diff
@@ -0,0 +1,57 @@
+const _ = require('lodash')
+const nodemailer = require('nodemailer')
+const Promise = require('bluebird')
+
+/**
+ * @hidden true
+ */
+
+const getTransport = async botId => {
+  const config = await bp.config.getModuleConfigForBot('skill-email', botId)
+
+  if (!config || !config.transportConnectionString || config.transportConnectionString === '<<change me>>') {
+    throw new Error(
+      `You must configure the "skill-email" module with valid SMTP credentials to send emails. Please see 'global/config/skill-email.json'.`
+    )
+  }
+
+  return config.transportConnectionString
+}
+
+const extractTextFromPreview = text => text.replace('(missing translation) ', '').replace(/([A-Z0-9_ -]+: )/gi, '')
+
+const sendEmail = async () => {
+  try {
+    const transport = await getTransport(event.botId)
+    const transporter = nodemailer.createTransport(transport)
+
+    const language = _.get(event, 'state.user.language', undefined)
+    const defaultLanguage = (await bp.bots.getBotById(event.botId)).defaultLanguage
+
+    const subjectEl = await bp.cms.getContentElement(event.botId, args.subjectElement, language)
+    const contentEl = await bp.cms.getContentElement(event.botId, args.contentElement, language)
+    const subject = subjectEl.previews[language] || subjectEl.previews[defaultLanguage]
+    const content = contentEl.previews[language] || contentEl.previews[defaultLanguage]
+
+    const mailOptions = {
+      from: args.fromAddress,
+      to: args.toAddress,
+      subject: extractTextFromPreview(subject),
+      text: extractTextFromPreview(content),
+      html: extractTextFromPreview(content)
+    }
+
+    await Promise.fromCallback(cb => transporter.sendMail(mailOptions, cb))
+    event.state.temp.success = true
+  } catch (error) {
+    event.state.temp.success = false
+    event.state.__error = {
+      type: 'action-execution',
+      stacktrace: error.stack,
+      actionName: 'skill-email/send_email',
+      actionArgs: args
+    }
```

Comment:
That should not be necessary, what do you think of just throwing the exception and letting the dialog engine handle it ? You could set event.state.temp.onErrorFlowTo if the goal was to continue to a specific node after the exception

Reply:
the problem I faced is that state is not persisted if there's an error.  didn't want to change the core to implement this, but the proper fix should be to save state changes even when there's an error

Reply:
The state should be persisted if there is an error.  I just tested it to be extra sure. In what situation does it not work? 

![image](https://user-images.githubusercontent.com/42552874/65283891-29dac280-db06-11e9-88a9-75ecfc13ec66.png)